File: archived_systems/4runr-agents/scripts/monitor_logs.py
#!/usr/bin/env python3
"""
Log Monitoring Script

This script monitors the logs of the 4Runr Multi-Agent System.
It checks for errors and warnings in the logs and sends alerts if necessary.
"""

import os
import sys
import re
import time
import logging
from datetime import datetime, timedelta

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger('log-monitor')

# Constants
LOG_DIR = "logs"
ERROR_PATTERN = re.compile(r'error|exception|fail|critical', re.IGNORECASE)
WARNING_PATTERN = re.compile(r'warning|warn', re.IGNORECASE)

def get_log_files():
    """Get a list of log files in the logs directory"""
    if not os.path.isdir(LOG_DIR):
        logger.error(f"Logs directory not found: {LOG_DIR}")
        return []
    
    return [os.path.join(LOG_DIR, f) for f in os.listdir(LOG_DIR) if f.endswith(".log")]

def check_log_file(log_file, since=None):
    """Check a log file for errors and warnings"""
    if not os.path.isfile(log_file):
        logger.error(f"Log file not found: {log_file}")
        return 0, 0
    
    # Get the file modification time
    mtime = os.path.getmtime(log_file)
    if since and mtime < since.timestamp():
        # File hasn't been modified since the specified time
        return 0, 0
    
    error_count = 0
    warning_count = 0
    
    try:
        with open(log_file, 'r') as f:
            for line in f:
                if ERROR_PATTERN.search(line):
                    error_count += 1
                elif WARNING_PATTERN.search(line):
                    warning_count += 1
    except Exception as e:
        logger.error(f"Error reading log file {log_file}: {e}")
    
    return error_count, warning_count

def monitor_logs(interval=60, alert_threshold=5):
    """Monitor logs for errors and warnings"""
    logger.info(f"Starting log monitoring (interval: {interval}s, alert threshold: {alert_threshold} errors)")
    
    last_check = datetime.now()
    
    while True:
        # Get the list of log files
        log_files = get_log_files()
        
        if not log_files:
            logger.warning("No log files found")
            time.sleep(interval)
            continue
        
        # Check each log file
        total_errors = 0
        total_warnings = 0
        
        for log_file in log_files:
            errors, warnings = check_log_file(log_file, since=last_check)
            total_errors += errors
            total_warnings += warnings
            
            if errors > 0 or warnings > 0:
                logger.info(f"{os.path.basename(log_file)}: {errors} errors, {warnings} warnings")
        
        # Send alerts if necessary
        if total_errors >= alert_threshold:
            logger.error(f"ALERT: {total_errors} errors found in logs")
            # In a real implementation, this would send an email or other alert
        
        # Update the last check time
        last_check = datetime.now()
        
        # Sleep until the next check
        time.sleep(interval)

def main():
    """Main function to run the log monitor"""
    logger.info("Starting log monitor...")
    
    # Parse command-line arguments
    interval = 60  # Default: check every 60 seconds
    alert_threshold = 5  # Default: alert if 5 or more errors
    
    if len(sys.argv) > 1:
        try:
            interval = int(sys.argv[1])
        except ValueError:
            logger.error(f"Invalid interval: {sys.argv[1]}")
            return 1
    
    if len(sys.argv) > 2:
        try:
            alert_threshold = int(sys.argv[2])
        except ValueError:
            logger.error(f"Invalid alert threshold: {sys.argv[2]}")
            return 1
    
    try:
        monitor_logs(interval, alert_threshold)
    except KeyboardInterrupt:
        logger.info("Log monitor stopped by user")
    except Exception as e:
        logger.error(f"Error in log monitor: {e}")
        return 1
    
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: archived_systems/4runr-lead-system/scripts/monitor_logs.js
#!/usr/bin/env node
/**
 * Log Monitoring Script
 * 
 * This script monitors the logs of the 4Runr Lead System.
 * It checks for errors and warnings in the logs and sends alerts if necessary.
 */

const fs = require('fs');
const path = require('path');
const readline = require('readline');
const { promisify } = require('util');

// Constants
const LOG_DIR = path.join(__dirname, '..', 'logs');
const ERROR_PATTERN = /error|exception|fail|critical/i;
const WARNING_PATTERN = /warning|warn/i;
const DEFAULT_INTERVAL = 60; // seconds
const DEFAULT_ALERT_THRESHOLD = 5;

// Promisify fs functions
const readdir = promisify(fs.readdir);
const stat = promisify(fs.stat);

/**
 * Get a list of log files in the logs directory
 * @returns {Promise<string[]>} Array of log file paths
 */
async function getLogFiles() {
  try {
    // Check if the logs directory exists
    if (!fs.existsSync(LOG_DIR)) {
      console.error(`Logs directory not found: ${LOG_DIR}`);
      return [];
    }
    
    // Get all files in the logs directory
    const files = await readdir(LOG_DIR);
    
    // Filter for log files
    return files
      .filter(file => file.endsWith('.log'))
      .map(file => path.join(LOG_DIR, file));
  } catch (error) {
    console.error(`Error getting log files: ${error.message}`);
    return [];
  }
}

/**
 * Check a log file for errors and warnings
 * @param {string} logFile - Path to the log file
 * @param {Date} [since] - Only check entries since this time
 * @returns {Promise<{errors: number, warnings: number}>} Count of errors and warnings
 */
async function checkLogFile(logFile, since = null) {
  try {
    // Check if the file exists
    if (!fs.existsSync(logFile)) {
      console.error(`Log file not found: ${logFile}`);
      return { errors: 0, warnings: 0 };
    }
    
    // Get the file modification time
    const stats = await stat(logFile);
    if (since && stats.mtime < since) {
      // File hasn't been modified since the specified time
      return { errors: 0, warnings: 0 };
    }
    
    // Read the file line by line
    const fileStream = fs.createReadStream(logFile);
    const rl = readline.createInterface({
      input: fileStream,
      crlfDelay: Infinity
    });
    
    let errorCount = 0;
    let warningCount = 0;
    
    for await (const line of rl) {
      if (ERROR_PATTERN.test(line)) {
        errorCount++;
      } else if (WARNING_PATTERN.test(line)) {
        warningCount++;
      }
    }
    
    return { errors: errorCount, warnings: warningCount };
  } catch (error) {
    console.error(`Error checking log file ${logFile}: ${error.message}`);
    return { errors: 0, warnings: 0 };
  }
}

/**
 * Monitor logs for errors and warnings
 * @param {number} interval - Check interval in seconds
 * @param {number} alertThreshold - Number of errors to trigger an alert
 */
async function monitorLogs(interval = DEFAULT_INTERVAL, alertThreshold = DEFAULT_ALERT_THRESHOLD) {
  console.log(`Starting log monitoring (interval: ${interval}s, alert threshold: ${alertThreshold} errors)`);
  
  let lastCheck = new Date();
  
  while (true) {
    // Get the list of log files
    const logFiles = await getLogFiles();
    
    if (logFiles.length === 0) {
      console.warn('No log files found');
      await new Promise(resolve => setTimeout(resolve, interval * 1000));
      continue;
    }
    
    // Check each log file
    let totalErrors = 0;
    let totalWarnings = 0;
    
    for (const logFile of logFiles) {
      const { errors, warnings } = await checkLogFile(logFile, lastCheck);
      totalErrors += errors;
      totalWarnings += warnings;
      
      if (errors > 0 || warnings > 0) {
        console.log(`${path.basename(logFile)}: ${errors} errors, ${warnings} warnings`);
      }
    }
    
    // Send alerts if necessary
    if (totalErrors >= alertThreshold) {
      console.error(`ALERT: ${totalErrors} errors found in logs`);
      // In a real implementation, this would send an email or other alert
    }
    
    // Update the last check time
    lastCheck = new Date();
    
    // Sleep until the next check
    await new Promise(resolve => setTimeout(resolve, interval * 1000));
  }
}

/**
 * Main function to run the log monitor
 */
async function main() {
  console.log('Starting log monitor...');
  
  // Parse command-line arguments
  const args = process.argv.slice(2);
  const interval = args[0] ? parseInt(args[0], 10) : DEFAULT_INTERVAL;
  const alertThreshold = args[1] ? parseInt(args[1], 10) : DEFAULT_ALERT_THRESHOLD;
  
  if (isNaN(interval)) {
    console.error(`Invalid interval: ${args[0]}`);
    process.exit(1);
  }
  
  if (isNaN(alertThreshold)) {
    console.error(`Invalid alert threshold: ${args[1]}`);
    process.exit(1);
  }
  
  try {
    await monitorLogs(interval, alertThreshold);
  } catch (error) {
    console.error(`Error in log monitor: ${error.message}`);
    process.exit(1);
  }
}

// Run the main function
main().catch(error => {
  console.error(`Unhandled error: ${error.message}`);
  process.exit(1);
});